lib/realtime/tenants/connect.ex

```diff
@@ -78,6 +87,21 @@ defmodule Realtime.Tenants.Connect do
   end
 
   @impl GenServer
+  def handle_continue(:setup_connected_users, state) do
+    %{check_connected_user_interval: check_connected_user_interval} = state
+    Process.send_after(self(), :check_connected_users, check_connected_user_interval)
+    {:noreply, state}
+  end
+
+  @impl GenServer
+  def handle_info(:check_connected_users, %{tenant_id: tenant_id} = state) do
```

Comment:
Needs to continue checking on the interval.

Reply:
🤦‍♂️ ...

Reply:
fixed 👍

Reply:
There are two race conditions in this code:

1. What happens if someone attempts to connect at the time you are disconnecting? The process will die and the user will likely get a connection error.

2. You may also run this check at the exact moment someone is reconnecting, so you have zero clients, but then you will have to immediately start the process again. This may not be a major issue though. It just means you will shut down a process to immediately start again.

I am not well versed to fully validate a solution but here are ideas for each.

## Problem 1

1. Instead of immediately returning {:stop, :normal}, instead, remove yourself from the registry

2. Then, in 5 seconds, if someone is connected directly to you, you add yourself back. This may be tricky because you use `syn` to measure how many connections you have, and you may need other mechanisms (like monitoring clients)

## Problem 2

1. Instead of immediately disconnecting, ping every 30 seconds. In the first 30 seconds, if there is nobody connected, you will set a  "will_disconnect" to true for the next ping. Otherwise the "will_disconnect" is false. If in the second ping "will_disconnect" is true, then you stop. This can be trivially done by changing the message itself to `{:check_connected_users, will_disconnect}`. But, this is not a major issue, it is free to choose code simplicity.

Reply:
Made the change to have that double check. 

Also took the opportunity to kill the db_conn which will trigger the killing of the process but it feels that this should be the way right? 

But at the same time feels like another race condition 😅

Reply:
Sorry, my comment was not the clearest. You need to implement both solutions to solve both problems. The first solution (for Problem 1) is described in 2 steps.

Reply:
Altogether, you need to implement 3 steps:

1. Instead of immediately disconnecting, ping every 30 seconds. In the first 30 seconds, if there is nobody connected, you will set a "will_disconnect" to true for the next ping. Otherwise the "will_disconnect" is false. If in the second ping "will_disconnect" is true, then you stop. This can be trivially done by changing the message itself to {:check_connected_users, will_disconnect}.

2. Then, instead of immediately returning {:stop, :normal}, instead, remove yourself from the registry

3. Then, in 5 seconds, if someone is connected directly to you, you add yourself back. This may be tricky because you use syn to measure how many connections you have, and you may need other mechanisms (like monitoring clients)

Reply:
Oh got it!

I wonder if I will be able to do it with https://hexdocs.pm/syn/syn.html#unregister/2 🤔 

As for tracking connected clients, do you think the state of this process could be a good spot? It's done during the join process and potentially a good spot

Reply:
An idea...

Check the count in :syn every 10 seconds, keep a list of those counts with the latest 6 (one minute), if all of those 6 are zero then kill the db connection. 

Note:

Clients are not connected directly to anything really. They get their messages to deliver via PubSub.

Just want to make sure we're not duplicating (or hand rolling) registries where we don't need to.

A "tenant" is registered via :syn and we use that as the authority for who is connected to which tenant. We use this for for our custom PubSub dispatcher for the database records. I guess we're already kind of duplicating registries in this case. Even more reason not to create another probably.

Reply:
Also note:

There is the `Extensions.PostgresCdcRls.SubscriptionManager` which replicates this data to a subscription table in the tenant database.

It also does the "stop this connection after a period of no connected users" thing:

https://github.com/supabase/realtime/blob/7bda695d927546762b14bf15185030da9117683a/lib/extensions/postgres_cdc_rls/subscription_manager.ex#L190

Which we can pull out if we get the `Connect` module managing all of these procs (todo later).

Reply:
Need to handle processes in state there because we need to serialize adding and removing subscriptions to the database.  

But in this "health check" process we don't really need that complexity at the moment.

Reply:
Sorry (rambling) I guess yeah you could just call the `Connect` process to register a Channel process with in and in that you would monitor the channel process for when it goes down.

Only thing here is that we need to ensure it can handle lots of connects/disconnects (assume one million).

Reply:
I think that you are right in thinking that at the moment it might be too much complexity as this process should only care about DB connection and shouldn't manage channels and if we add state that reacts to channels lifecycles we would open that Pandora box

To recap, for this module what we have:
- Check if no one is using channels associated with this tenant (counter already exists)
- If no one is using it, we disconnect the DB Connection
- If no one is using and db connection is dead, removes the Process from the Registry by killing it
- We won't be able to be very graceful when new users join since we're using syn which needs time to replicate across the network

In my eyes we have two options that can avoid full channel management and cover concerns mentioned:

## potential solution 1
after checking X times that no one is connected which trigger the unregister from syn (process still alive). Instead of controlling all connected users we can use the state of this process to track if "someone" is trying to reconnect by setting a single flag. If "is_connecting" is true register again. If flag is false kill the db conn process and Connect process after X seconds

- less sync (aka more eventually consistent)
- more processes being killed / started up
- syn might lead to conflicts but we can use conflict resolution already implemented


## potential solution 2
This last step IMO is the tricky one since it can die at the wrong time so what we can do is take advantage of the fact that we can associate metadata with syn and instead of killing the Connection process we can change its metadata to db_conn nil. With this change, upon new join, if we have db_conn metadata nil we know that we need to reconnect and that flow will become more "sync" since we are using only this process mailbox to receice and handle it.

- more sync
- less process changes in the system
- this process will become more of a bottleneck

Reply:
opted to go with a simpler solution for now. ended up using the idea of buckets so we can better control the shutdown of the process (can also be useful in the future for more complex behaviours) but the process unregister is done on process shutdown so that way we keep the behaviour consistent with other parts of the code.

later we can revisit this code as this module will need more elements as we bring new features

https://github.com/supabase/realtime/pull/701/commits/486a8bc8c53b1de1fb66ef09ce994bb656ba038f